### PR TITLE
Feature/zenko 3368/add chain backend

### PR DESCRIPTION
**Pull request by alexanderchan-scality**
* reorganized `in_memory` backend files
* adds auth `ChainBackend`:
  * allows for authentication to be done using multiple auth backends or vault clients
  * authenticates in order of client/backend list provided until one succeeds

### CODE REVIEW

File: lib/auth/backends/ChainBackend.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    /*
+     * apply task to all clients
+     */
+    _forEachClient(task, cb) {
+        async.map(this._clients, task, cb);
+    }
+
+    verifySignatureV2(stringToSign, signatureFromRequest, accessKey, options, callback) {
+        this._tryEachClient((client, done) => {
+            if (typeof client.verifySignatureV2 !== 'function') {
```
Comment (rachedbenmustapha) [inline lib/auth/backends/ChainBackend.js]: This is a pretty expensive check to do at each invocation. I’d check for all backends in the chain in the constructor and raise an exception if any of the functions is missing (which should happen close to never).

File: lib/auth/backends/ChainBackend.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            }
+            return callback(null, {
+                message: {
+                    body: ChainBackend._mergePolicies(res),
+                },
+            });
+        });
+    }
+
+    healthCheck(reqUid, callback) {
```
Comment (rachedbenmustapha) [inline lib/auth/backends/ChainBackend.js]: Is the difference in case intentional? (healthCheck vs healthcheck)

**Review by rachedbenmustapha — APPROVED** [review]

File: lib/auth/backends/ChainBackend.js (hunk 2 above)
Comment (alexanderchan-scality) [reply lib/auth/backends/ChainBackend.js]: good catch. this is a typo will fix, as there is a fallback, the tests I have done were false positive passes

**Review by naren-rajendran — APPROVED** [review]

File: lib/auth/backends/ChainBackend.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        this._forEachClient((client, done) => {
+            if (typeof client.healthcheck !== 'function') {
+                return done(null, { code: 200, message: 'OK' });
+            }
+            return client.healthcheck(reqUid, done);
+        }, (err, res) => {
+            if (err) {
+                return callback(err);
+            }
+            return callback(null, res);
```
Comment (rachedbenmustapha) [inline lib/auth/backends/ChainBackend.js]: Seems fine in that if any backend fails the check, an error will be returned, but doesn't this change the shape of `res`?

File: lib/auth/backends/ChainBackend.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+    static _mergeObjects(objectResponses) {
+        return objectResponses.reduce(
+            (retObj, resObj) => Object.assign(retObj, resObj.message.body),
+            {});
+    }
+
+    getCanonicalIds(emailAddresses, options, callback) {
+        this._forEachClient((client, done) => {
+            if (typeof client.getCanonicalIds !== 'function') {
```
Comment (rachedbenmustapha) [inline lib/auth/backends/ChainBackend.js]: This and `getEmailAddresses` would also benefit from being checked once only. If it's expected that one backend doesn't implement them (which I don't think is the case), then just checking truthiness would be enough.

**Review by rachedbenmustapha — COMMENTED** [review]
We don't have time to perform performance testing unfortunately, but this change will definitely need to be evaluated for performance.

File: lib/auth/backends/ChainBackend.js (hunk 3 above)
Comment (alexanderchan-scality) [reply lib/auth/backends/ChainBackend.js]: What I had in mind was to just returning the error in the case of that  the healthcheck fails
after taking a step back and having a better look at this, res is even more important in the case of an error.

File: lib/auth/backends/ChainBackend.js (hunk 4 above)
Comment (alexanderchan-scality) [reply lib/auth/backends/ChainBackend.js]: yes. You're right. If a client/backend doesn't implement all these methods, it should be assumed to be invalid.

File: lib/auth/backends/ChainBackend.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            }
+            return callback(null, {
+                message: {
+                    body: ChainBackend._mergePolicies(res),
+                },
+            });
+        });
+    }
+
+    healthcheck(reqUid, callback) {
```
Comment (alexanderchan-scality) [inline lib/auth/backends/ChainBackend.js:167]: @naren-scality @RachedBenMustapha updated the healthcheck to test all clients and not exit on early error.

**Review by rachedbenmustapha — APPROVED** [review]

**Pull request merged**